src/Functions/idna.cpp

```diff
@@ -0,0 +1,202 @@
+#include "config.h"
+
+#if USE_IDNA
+
+#include <Columns/ColumnString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionStringToString.h>
+
+#ifdef __clang__
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wnewline-eof"
+#endif
+#    include <ada/idna/to_ascii.h>
+#    include <ada/idna/to_unicode.h>
+#    include <ada/idna/unicode_transcoding.h>
+#ifdef __clang__
+#    pragma clang diagnostic pop
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int NOT_IMPLEMENTED;
+}
+
+/// Implementation of
+/// - idnaEncode(), tryIdnaEncode() and idnaDecode(), see https://en.wikipedia.org/wiki/Internationalized_domain_name#ToASCII_and_ToUnicode
+///   and [3] https://www.unicode.org/reports/tr46/#ToUnicode
+
+enum class ErrorHandling
+{
+    Throw,  /// Throw exception
+    Empty   /// Return empty string
+};
+
+
+/// Translates a UTF-8 string (typically an Internationalized Domain Name for Applications, IDNA) to an ASCII-encoded equivalent. The
+/// encoding is performed per domain component and based on Punycode with ASCII Compatible Encoding (ACE) prefix "xn--".
+/// Example: "straße.münchen.de" --> "xn--strae-oqa.xn--mnchen-3ya.de"
+/// Note: doesn't do percent decoding. Doesn't trim tabs, spaces or control characters. Expects non-empty inputs.
+template <ErrorHandling error_handling>
+struct IdnaEncode
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const size_t rows = offsets.size();
+        res_data.reserve(data.size()); /// just a guess, assuming the input is all-ASCII
+        res_offsets.reserve(rows);
+
+        size_t prev_offset = 0;
+        std::string ascii;
+        for (size_t row = 0; row < rows; ++row)
+        {
+            const char * value = reinterpret_cast<const char *>(&data[prev_offset]);
+            const size_t value_length = offsets[row] - prev_offset - 1;
+
+            std::string_view value_view(value, value_length);
+            if (!value_view.empty()) /// to_ascii() expects non-empty input
```

Comment:
Shouldn't `idnaEncode` throw an exception for an empty input?

Reply:
Empty input is an edge case which is generally legal. `ada::idna::to_ascii()` handles empty inputs gracefully (it doesn't throw, it returns an empty output). The comment on l. 65 is only there because the [API docs](https://github.com/ada-url/idna/blob/main/include/ada/idna/to_ascii.h#L13) say so and I wanted to program against the API, not the implementation. In any case, it slightly simplifies the actual error handling (l. 68ff) which would otherwise be ambiguous in the empty string case.